File: paradygmaty/lista3/zad2/DHSetup.java
import java.util.Random;
import java.util.ArrayList;

public class DHSetup<T extends GF>{
    private T generator;
    public long p;
    
    public DHSetup() {
        this.p = T.get_p();

        ArrayList<Long> primes = new ArrayList<>();

        for (long i = 2; i * i <= (p - 1); ++i) {
            if ((p - 1) % i == 0) {
                primes.add(i);
                if (i != (p - 1) / i) {
                    primes.add((p - 1) / i);
                }
            }
        }

        setGenerator(primes);
    }

    public void setGenerator(ArrayList<Long> primes) {
        Random random = new Random();
        long potentialGenerator;
        do {
            potentialGenerator = random.nextLong() % (p - 1) + 1;
        } while (!isValidGenerator(potentialGenerator, p - 1, primes));
        
        generator = (T) T.newInstance(potentialGenerator);
    }
    
    public T getGenerator() {
        return generator;
    }

    public T power(T a, long b) {
        T result = (T) T.newInstance(1);
        while (b > 0) {
            if ((b & 1) == 1) {
                result = (T) T.mul(result, a);
            }
            a = (T) T.mul(a, a);
            b >>= 1;
        }
        return result;
    }
    
    private boolean isValidGenerator(long a, long pMinusOne, ArrayList<Long> primes) {
        for (Long q : primes)
        {
            if (pMinusOne % q == 0) {
                if (T.eq(power((T) T.newInstance(a), q), (T) T.newInstance(1))) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: paradygmaty/lista3/zad3/dhsetup.py
import math
import random

class dhsetup:
    def __init__(self, T):
        self.T = T
        self.generator = self.generate_generator()

    def is_generator(self, a, p, primes):
        for q in primes:
            if self.power(a, (p - 1) // q) == 1:
                return False
        return True

    def generate_generator(self):
        p = self.T.get_characteristic()

        primes = []
        for i in range(2, int(math.sqrt(p - 1)) + 1):
            if (p - 1) % i == 0:
                primes.append(i)
                if i != (p - 1) // i:
                    primes.append((p - 1) // i)

        while True:
            gen_value = random.randint(2, p - 1)
            a = self.T(gen_value)
            if self.is_generator(a, p, primes):
                return a

    def get_generator(self):
        return self.generator

    def power(self, a, b):
        result = self.T(1)
        while b > 0:
            if b % 2 == 1:
                result = result * a
            a = a * a
            b //= 2
        return result
